### DISCUSSION TITLE
💡 react-router (v6.4+) `loader`  API with react-query

### DISCUSSION THREAD

**Discussion by cliffordfajardo** [TanStack/query#3655] [Q&A] [answered]
### _**🌕 The `loader` API for `react-router` is in pre-release state as this time**_ 

Hey RQ community!
I created this thread so folks can share solutions / ideas / questions when adopting react-router v6.4+ (_the one with `loaders` API_).
V6.4 may be particularly interesting update to all users of react-router, since it brings concepts from [Remix (full React framework)](https://remix.run/) to our client side rendered React apps. 

Potentially this means, if you start adopting the newer `react-router` API's, it could significantly help you transition to adopting Remix, as Remix uses the v6.4 react-router APIs already.

## Why this might be important to `react-query` users?
1. Most of us using `react-query` are also using `react-router
2. Some folks may have an interest in adopting [Remix](https://remix.run/) in the near future, which uses the same `loader` API
3. **Even if you don't adopt Remix,** the `loaders` API can help you avoid request waterfall chains, since its able to initiate in parallel all the API requests in app. (For more context see the video link below)

<a href="https://youtu.be/Ck-e3hd3pKw?t=9874" target="_blank">
<img width="944" alt="CleanShot 2022-05-29 at 10 11 10@2x" src="https://user-images.githubusercontent.com/6743796/170882809-a9435869-1642-43e6-a978-62189196a9db.png">
</a>

With that said, there isn't too much information about using`react-query` and the new `react-router` API together (_pros/cons/best practices etc_), hence the creation of this thread to bring awareness / spark conversations / get people cross linking and sharing resources.

##  New react-router `loader` API Summary
The new `react-router` `loader` API allows you define a "loader" function to provide data to the route element before it renders.

```tsx
// https://beta.reactrouter.com/en/remixing/route/loader
// 🚧 This feature only works if using `DataBrowserRouter`
<DataBrowserRouter>
  <Route
    element={<Teams />}
    path="teams"
    loader={async () => {
      return fakeDb.from("teams").select("*");
    }}
  >
    <Route
      element={<Team />}
      path=":teamId"
      loader={async ({ params }) => {
        return fetch(`/api/teams/${params.teamId}.json`);
      }}
    />
  </Route>
</DataBrowserRouter>
```

**Example `create-react-app`**
I created a simple `create-react-app` showing the loader API. 
Im going to continue playing around with new `loader` API more and reporting back here.

- https://github.com/cliffordfajardo/react-query-with-react-router-loader
![CleanShot 2022-05-29 at 12 36 45@2x](https://user-images.githubusercontent.com/6743796/170888553-5f2d7ea6-8473-4350-bff1-765bacfac6fb.png)

## References
- [Example `create-react-app` using `react-query` with new `loader` API](https://github.com/cliffordfajardo/react-query-with-react-router-loader)
- [Remixing React Router blog post](https://remix.run/blog/remixing-react-router)
- [YouTube - Ryan Florence talking about react-router loader API](https://www.youtube.com/watch?t=9275&v=Ck-e3hd3pKw&feature=youtu.be)
- [Beta React Router docs](https://beta.reactrouter.com/en/remixing/route/loader)
- [Twitter Discussion thread that sparked this Github Discussion](https://twitter.com/cliffordfajard0/status/1530970753855655936)

<details>

<summary>Tangential Aside: My teammates & I we are ecstatic about Remix for a few reason(s):</summary>

- Working in a fullstack typescript based React stack for our web UI
   - no context switching between languages
   - use the web platform / APIs for both the UI and frontend web server
- If you're using `react-router` the same `react-router` API is used in Remix
- Remix allows us to SSR our app easily
- Remix as a framework has lots of good conventions, which result in out of the box "gaurd rails" on the developer experience, performance and user experience side
</details>

**Comment by ryanflorence** [comment]
I'll edit this answer more later, don't have much time, but wanted to quickly communicate my thoughts. I've had RQ in mind the whole time while we did this work. Admittedly, I've never used RQ in earnest, so I don't know the intricacies of it's behavior like background fetching, etc.

For React Router, the general abstraction is about three things: initiating, mutating, and reading. It maps to React Router like so:

```
loader -> initiate
action -> mutate
component -> read
```

React Router's behavior is pretty simple to explain (though pretty intricate in the implementation)

- Location changes
- New/changing loaders for the next location are called and awaited before changing screens
- Once all loaders are done, the page changes over to the new location

Next, when an action is called from a form or imperative submission:

- the leaf route action is called and awaited
- the loaders on the page will be revalidated and awaited
- the screen updates once all of them are done

Along the way, Remix keeps track of loading and error states.

For React Query (with my limited experience with it) I think it looks like this:

```
loader -> fetchQuery (need to await it)
action -> invalidateQueries
component -> useQuery
```

I don't know if `fetchQuery` throws on errors or catches and returns. If it returns the error, you'll want to throw it in the loader.

There is nothing magic about `useLoaderData`, it doesn't do anything, it simply provides what was returned from the loader. In this case it's irrelevant because what you want is inside of RQ's cache not React Router's state.

That's pretty much it!

**Reply by cliffordfajardo** [reply]
Thanks @ryanflorence 

> For React Query (with my limited experience with it) I think it looks like this:
> loader -> fetchQuery (need to await it)
> action -> invalidateQueries
> component -> useQuery

Yeah, this looks like it works  🙌 
![CleanShot 2022-05-29 at 12 55 04@2x](https://user-images.githubusercontent.com/6743796/170889180-369a4d61-ff32-4e51-8fe9-d6cbd44e252d.png)


--------
I noticed in the `loader` function examples from the docs, we are not using `await` keyword; is this because the `loader` is `fetch` API aware and does the unwrapping of the data for us?

```tsx
// https://beta.reactrouter.com/en/remixing/route/loader
// 🚧 This feature only works if using `DataBrowserRouter`
<DataBrowserRouter>
  <Route
    element={<Teams />}
    path="teams"
    loader={async () => {
      //❓ no await here, so will react-router call just check if the promise's state is resolved?
      return fakeDb.from("teams").select("*");
    }}
  >
    <Route
      element={<Team />}
      path=":teamId"
      loader={async ({ params }) => {
        // ❓under the hood is the loader,  `fetch` API-aware to unwrap the `.json` data?
        return fetch(`/api/teams/${params.teamId}.json`);
      }}
    />
  </Route>
</DataBrowserRouter>
```




CC @TkDodo @aaronksaunders @Temzasse - thought you might be curious too 😄

**Reply by ryanflorence** [reply]
Just have to return a promise, that's all. React Router is aware of web fetch `Response` objects, so yes, you can simply return a fetch and it'll unwrap it, or just return whatever you want.

```js
// aware of responses, will unwrap with .json()
return new Response("hey!");

// fetch returns one of those, so it's kinda slick
return fetch("...");

// but really just return whatever you want
return { whatever: "you want" };

// it should be a promise that doesn't resolve until your data resolves:
return fetchQuery(...);
// not
return prefetchQuery(...);
```

**Comment by TkDodo** [comment]
As Ryan mentioned in his talk, if you're already using react-query, the loader api from react-router gives you a great place to start prefetching. Off the top of my head, there are two things I would keep an eye on:

- `queryClient.fetchQuery` returns a Promise with data and throws errors,`queryClient.prefetchQuery` also returns a Promise, but it doesn't contain data (`Promise<void>`) and it doesn't throw errors. Otherwise, they are the same (prefetchQuery calls fetchQuery internally). So if you want to use `useLoaderData` to read data for some reason over `useQuery`, you would need `fetchQuery`. If the loader expects an error to be thrown for error handling - also `fetchQuery`.
- I would strongly consider passing a `staleTime` to fetchQuery / prefetchQuery in the loader to avoid unnecessary calls to the queryFn if the data is still considered fresh. An underused tool is also `queryClient.setQueryDefaults(['todos'], { staleTime: 10000 })`, because it will set the default for this key (fuzzily) for wherever it is used.

**Comment by cliffordfajardo** [comment]
TK Dodo published some "work in progress" example code here using react-router 6.4+ with react-query ✨ 
- WIP Code: https://github.com/TanStack/query/pull/4058
- Tweet: https://twitter.com/TkDodo/status/1560971584642420736

**Comment by TkDodo** [comment, marked as answer]
I've merged the example and written a blogpost about it:
- https://tkdodo.eu/blog/react-query-meets-react-router
- https://tanstack.com/query/v4/docs/examples/react/react-router

**Reply by Toshinaki** [reply]
Thank you for the blog.
It has been almost a year since the post, any changes should we aware?
Such a great combination for router and query, though a pity not mentioned much…

And there're several things I'd like to make sure I understand it correctly:
1. if the loader is a sync function, then after click a link, react router will call the function and wait until the result is ready, then the component renders. is it right?
2. if it's true, does that means if the function is heavy, the route won't change until the function ends? or the suspense component shows up until function returns?
3. if the loader is an async function, no `await` is needed, router waits for the result is returned, then the component renders? or same with above, the suspense component renders?

**Reply by Toshinaki** [reply]
4. when used with react query, the component won't load if the query not end right? else if read data in the component through useQuery, since the query inside loader is not ended, the useQuery won't start a new fetch, it just wait the loader ends and read from cache right?

**Reply by TkDodo** [reply]
1. yes
2. it won't render, but you can `useNavigation` from `react-router-dom` to render an indicator during page transitions. The example does that. This is why I recommend [ensureQuery](https://tanstack.com/query/v4/docs/reference/QueryClient#queryclientensurequerydata), because it will resolve immediately when any data is in the cache already.
3. the loader just awaits the promise that your return. async/await is syntactic sugar for chaining promises with `.then`. If you don't await, it returns earlier, but data is not there. You need to understand what async/await does: https://tkdodo.eu/blog/about-async-functions

**Reply by frassinier** [reply]
@TkDodo is there any chance that react-query cache won't be reconciled, once the request is done, if we are using react-router v6 without any loader/action mechanism, but instead, just the Component scope (`useTodos()` hook for example)? 
Is there a refetch and mount/unmount while navigating as far as you know?
I'm puzzled by some issues in production, and only in production mode, where stale requests are sent but do not trigger any changes in my component!

**Reply by TkDodo** [reply]
I'm not sure what you mean exactly. Please show a reproduction of your problem.

**Reply by frassinier** [reply]
My question was mainly about "Are these changes to use loader/action mandatory or convenient reasons only?". 

I can't reproduce except on specific env but I can see sometimes the Axios query being processed and when I output data in `useInfiniteQuery` `select()` or `onSuccess()`, the value comes from the cache and differs from what Axios got, so I was wondering if it can be related to ReactRouter v6 and the way it's recommended to use loader and action.

**Reply by TkDodo** [reply]
> "Are these changes to use loader/action mandatory or convenient reasons only?

loaders are an optional feature that was added in React Router 6.4. You don't "have to" use it.

**Reply by kevinpruett-prime** [reply]
@TkDodo I'm not sure if this question is appropriate here, but I am curious about the techniques you've outlined above. Specifically, why do you prefer to use `useQuery` in the component as opposed to `useLoaderData`?

Perhaps I'm misunderstanding react router's `loaders/useLoaderData` but could you explain the reason for opting for `useQuery` instead of `useLoaderData` when the loader could pass the data through and could be consumed with the `userLoaderData` hook?

Also, do you recommend a good pattern for error handling throughout this lifecycle? (e.g. If a query fails within the `loader` definition, what should you return? should you throw immediately in the loader or somehow pass this through to your component?)

Thanks

**Reply by TkDodo** [reply]
> why do you prefer to use useQuery in the component as opposed to useLoaderData?

because it will give you data synchronization and background refetches instead of just getting data from the loader once: https://tkdodo.eu/blog/react-query-meets-react-router
